src/v/cloud_storage/remote_segment_index.h

```diff
@@ -0,0 +1,1139 @@
+        _data.append(reinterpret_cast<uint8_t*>(&bits1), sizeof(bits1));
+    }
+
+    void _pack6(const row_t& input) {
+        uint64_t bits0 = 0;
+        uint32_t bits1 = 0;
+        bits0 |= static_cast<uint64_t>((input[0] & 0x3F));
+        bits0 |= static_cast<uint64_t>((input[1] & 0x3F) << 6);
+        bits0 |= static_cast<uint64_t>((input[2] & 0x3F) << 12);
+        bits0 |= static_cast<uint64_t>((input[3] & 0x3F) << 18);
+        bits0 |= static_cast<uint64_t>((input[4] & 0x3F) << 24);
+        bits0 |= static_cast<uint64_t>((input[5] & 0x3F) << 30);
+        bits0 |= static_cast<uint64_t>((input[6] & 0x3F) << 36);
+        bits0 |= static_cast<uint64_t>((input[7] & 0x3F) << 42);
+        bits0 |= static_cast<uint64_t>((input[8] & 0x3F) << 48);
+        bits0 |= static_cast<uint64_t>((input[9] & 0x3F) << 54);
+        bits0 |= static_cast<uint64_t>((input[10] & 0x0F) << 60);
+        bits1 |= static_cast<uint32_t>((input[10] & 0x30) >> 4);
+        bits1 |= static_cast<uint32_t>((input[11] & 0x3F) << 2);
+        bits1 |= static_cast<uint32_t>((input[12] & 0x3F) << 8);
+        bits1 |= static_cast<uint32_t>((input[13] & 0x3F) << 14);
+        bits1 |= static_cast<uint32_t>((input[14] & 0x3F) << 20);
+        bits1 |= static_cast<uint32_t>((input[15] & 0x3F) << 26);
+        _data.append(reinterpret_cast<uint8_t*>(&bits0), sizeof(bits0));
+        _data.append(reinterpret_cast<uint8_t*>(&bits1), sizeof(bits1));
+    }
+
+    void _pack7(const row_t& input) {
+        uint64_t bits0 = 0;
+        uint32_t bits1 = 0;
+        uint16_t bits2 = 0;
+        bits0 |= static_cast<uint64_t>((input[0] & 0x7F));
+        bits0 |= static_cast<uint64_t>((input[1] & 0x7F) << 7);
+        bits0 |= static_cast<uint64_t>((input[2] & 0x7F) << 14);
+        bits0 |= static_cast<uint64_t>((input[3] & 0x7F) << 21);
+        bits0 |= static_cast<uint64_t>((input[4] & 0x7F) << 28);
+        bits0 |= static_cast<uint64_t>((input[5] & 0x7F) << 35);
+        bits0 |= static_cast<uint64_t>((input[6] & 0x7F) << 42);
+        bits0 |= static_cast<uint64_t>((input[7] & 0x7F) << 49);
+        bits0 |= static_cast<uint64_t>((input[8] & 0x7F) << 56);
+        bits0 |= static_cast<uint64_t>((input[9] & 0x01) << 63);
+        bits1 |= static_cast<uint32_t>((input[9] & 0x7E) >> 1);
+        bits1 |= static_cast<uint32_t>((input[10] & 0x7F) << 6);
+        bits1 |= static_cast<uint32_t>((input[11] & 0x7F) << 13);
+        bits1 |= static_cast<uint32_t>((input[12] & 0x7F) << 20);
+        bits1 |= static_cast<uint32_t>((input[13] & 0x1F) << 27);
+        bits2 |= static_cast<uint16_t>((input[13] & 0x60) >> 5);
+        bits2 |= static_cast<uint16_t>((input[14] & 0x7F) << 2);
+        bits2 |= static_cast<uint16_t>((input[15] & 0x7F) << 9);
+        _data.append(reinterpret_cast<uint8_t*>(&bits0), sizeof(bits0));
+        _data.append(reinterpret_cast<uint8_t*>(&bits1), sizeof(bits1));
+        _data.append(reinterpret_cast<uint8_t*>(&bits2), sizeof(bits2));
+    }
```

Comment:
Is there a reference implementation that would be worth linking to, or all of this is custom implementation of the encodings?

Reply:
❤️ 🎉